### ISSUE TITLE
Encountering a consistent indexing error on Matic "Found no transaction for event"

### ISSUE THREAD

**Issue by TravelingTechGuy** [graphprotocol/graph-node#2552]
1. This is a **bug report**
2. I'm trying to deploy my subgraph. It used the [ABI of the USDT contract](https://explorer-mainnet.maticvigil.com/address/0x7FFB3d637014488b63fb9858E279385685AFc1e2/contracts), and is supposed to index an entity whenever the `to` in the `Transfer` event is the address of our contract (i.e. whenever someone staked USDT in the contract).

This works well on Ethereum. On Matic (Polygon), we get the following fatal error:

```
"failed to process trigger: block #15167232 (0x7b2a…0d6b), transaction 4641fb604c8b6b9d515f57f364e8039fb600ed0471a7dcb7f16f2bd1314501f1: Found no transaction for event"
```

This is consistent, and reproducible. I tried adding a function that leaves the handler in case: the block number is 15167232, the block number is between 15167200 and 15167240 (thought a range would help), if `event.transaction` is empty - to no avail, this error still happens. My guess is this happens before my handler code, and therefore, not under my control.

My workaround is to add a higher `startBlock` number to the yaml, but the we lose tons of events.

My expected behavior is either a resolution that allows me to index the graph, or a way to bypass whatever problem triggers the fatal error.

More details:
1. The subgraph can be found at: https://thegraph.com/explorer/subgraph/travelingtechguy/cvipolygon
2. I can provide a link to the code, if needed. It has other entities that index well.

**Comment by TravelingTechGuy** [comment]
Update: I advanced the `startBlock` for the USDT contract past block 15167232, but it just blew up later with the same error:

```
failed to process trigger: block #15326592 (0x9641…dfa9), transaction 626d7374ca1862fe801fad8b57418fddd19f826b2d9797a014c5589da343985d: Found no transaction for event"
```

**Comment by azf20** [comment]
Thanks for reporting @TravelingTechGuy - can you please share a link to the code so we can replicate?
This looks like the same issue as #2353, but on a different network (Matic, not Celo) cc @tilacog and @leoyvens

**Comment by TravelingTechGuy** [comment]
Hi @azf20 - thanks for looking into this!

If you look at [this branch](https://github.com/coti-io/cvi-polygon-subgraph/tree/event_fix) - the USDT2.json is the USDT ABI I took from the Matic explorer. Change the `startBlock` in the yaml to a number below 15167232 and run the deployment. You can completely discard the other CVI entities. 

In fact, we've now separated them from the USDT contract, as they index well, and USDT blows up.

In short, I'd say the quickest recreation steps:

1. Create a new subgraph
2. Grab the UST ABI from the [explorer](https://explorer-mainnet.maticvigil.com/address/0x7FFB3d637014488b63fb9858E279385685AFc1e2/contracts)
3. Index the `Transfer` event
4. Start at `startBlock` close to one of the error ones.
5. Deploy

**Comment by azf20** [comment]
We have replicated this, it is possibly related to the fact that the `from` and `to` address for the transaction are both `0x0000000000000000000000000000000000000000`. Looking into it. Thanks for reporting!

e.g. here https://polygonscan.com/tx/0x626d7374ca1862fe801fad8b57418fddd19f826b2d9797a014c5589da343985d

**Comment by TravelingTechGuy** [comment]
Thanks for being on top of it @azf20!
I assume there's nothing I can do in my code to get around this, and that the error occurs prior to the handler being called?

I also wonder: a 0x to 0x transfer: it's not a burn per se - is that an indexing error, or some bot doing something?

**Comment by azf20** [comment]
It is a special case on Polygon, I think it relates to their bridge - the tokens aren't transferred from 0x to 0x, but the transaction itself is a `self` from 0x to 0x. We have a couple more instances of this, so hopefully we'll be able to isolate the issue

**Comment by TravelingTechGuy** [comment]
Thanks for the update @azf20 - i assumed it was something to do with Polygon, since we haven't run into it on mainnet.
In the meantime, is there anything I can do to guard against this in my code, or is it something that gets thrown before my code (handler) is even executed (i.e. I have no control of it)?

**Comment by azf20** [comment]
Yes I think this is throwing before the handler is initiated - @tilacog is investigating!

**Comment by azf20** [comment]
We have identified the issue, which relates to our Matic RPC not including all transactions in `eth_getBlockByHash` at some point, which was then cached in the database, resulting in the ongoing issue. This may relate to the new Bor requirement to specifically return bor logs. We are planning to clear the cache which should resolve the issue, as Matic RPCs are now returning the correct number of transactions for these blocks cc @leoyvens @tilacog

**Comment by azf20** [comment]
We cleared the cache, which has resolved this issue for the subgraph we were testing with - @TravelingTechGuy you should be able to restart this subgraph before the problematic block (15167232), and it should index successfully. I will close this issue, but please do re-open if you see a recurrence!

**Comment by azf20** [comment]
This recurred on the Decentraland subgraph:
```
Subgraph instance failed to run: failed to process trigger: block #16574080 (0x502d…2b95), transaction 384650eb47fd17640598f72eb3eeff356ab16a02ec3c5c33d3de818763a29081: Found no transaction for event, code: SubgraphSyncingFailure
```

**Comment by tilacog** [comment]
Since this error is solved by re-fetching the block, would it make sense to implement some retry mechanism whenever this error surfaces?

**Comment by miguelmota** [comment]
Started encountering this error as well and I'm not sure how to get around it. Any help is greatly appreciated

```bash
$curl -s --location --request POST https://api.thegraph.com/index-node/graphql --data-raw '{"query":"{ indexingStatusForPendingVersion(subgraphName: \"hop-protocol/hop-polygon\") { subgraph fatalError { message } nonFatalErrors {message } } }"}'
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16573056 (0x00d5…7cad), transaction 059189fea36971157989bf3c13b17743c9313df88f317b3fa68d59793c117c03: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmWEF3CEusZDBLwkM2ijUoFH8yuFbaFZcE2EAUpr8ykxbk"}}}
```

**Comment by leoyvens** [comment]
@tilacog The question is when do we know if we need to retry. Maybe if we double check the array of transactions received against `eth_getTransactionCount`?

**Comment by tilacog** [comment]
@miguelmota We've re-fetched the matic faulty block at the hosted service. Can you please confirm if the error is gone now?

**Comment by miguelmota** [comment]
@tilacog A subgraph reedeployment was made but the issue occurred again towards the end of the sync

https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending

![DeepinScreenshot_select-area_20210715151904](https://user-images.githubusercontent.com/168240/125865080-c9407c4e-287b-4643-a4d6-16c3cfb6c926.png)

**Comment by tilacog** [comment]
@miguelmota Thanks for the extra info.
I've re-fetched three faulty blocks from the hosted service, and this subgraph resumed syncing.

Please let us know if this incident happens again, and we'll try to fix the problem.

**Comment by miguelmota** [comment]
@tilacog looks to be fixed! Thanks!

**Comment by miguelmota** [comment]
@tilacog There's another failure here

https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?selected=logs

![DeepinScreenshot_select-area_20210720203855](https://user-images.githubusercontent.com/168240/126426985-72fb017b-0ce0-44ea-971a-b7acd2895207.png)

![DeepinScreenshot_select-area_20210720203947](https://user-images.githubusercontent.com/168240/126427052-28699885-4067-480e-9cb2-ae5096eb4e54.png)

**Comment by tilacog** [comment]
@miguelmota 
the subgraph has resumed syncing.

Thanks for the heads up!

**Comment by reddyismav** [comment]
@leoyvens did you happen to try what you suggested , eth_getTransactionCount and then comparing it with the array returned from getBlockByHash. ?

**Comment by reddyismav** [comment]
In the meanwhile m trying to figure out why the sync issue on the bor nodes is happening and why is it aways for state sync transactions. 
I ll keep you guys updated here.

**Comment by miguelmota** [comment]
@tilacog thanks again!

**Comment by nachomazzara** [comment]
Could it be related to reorgs?

**Comment by leoyvens** [comment]
@nachomazzara we get this also for reasonably old blocks, so I don't think so.

**Comment by nachomazzara** [comment]
> @nachomazzara we get this also for reasonably old blocks, so I don't think so.

Could the graph-node cache have the info of the reorged block and therefore couldn't find the transaction because it was reorged. In short, the tx is not there anymore

**Comment by miguelmota** [comment]
The polygon indexer is showing another sync failure

https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending

![DeepinScreenshot_select-area_20210726155002](https://user-images.githubusercontent.com/168240/127069047-3ddd017c-ccc8-43ff-8d5c-12ada0079666.png)

Apologies for being an annoyance with the same issue; let me know if there's a way I can trigger a block resync or handle the error. Thanks!

**Comment by azf20** [comment]
@nachomazzara the issue seems pretty specifically tied to 0x000 -> 0x000 transactions, rather than a more general re-org issue (if it was as you described, I think we would see it for other transactions, unless I have misunderstood your suggestion?)

Thanks for flagging @miguelmota, @tilacog can you take a look?

**Comment by miguelmota** [comment]
@tilacog any update on the sync failure on [subgraph](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending)?

**Comment by tilacog** [comment]
Hey @miguelmota, I'm following that subgraph closely and re-fetching faulty blocks from our storage. 
As of now, it has resumed syncing, but I'll keep an eye on it.

Unfortunately, we'll have to do this manually until this issue persists.

**Comment by azf20** [comment]
Hypothesis that this is caused by a race condition relating to use of block number instead of block hash for Bor transactions, which should be fixed in this PR: https://github.com/maticnetwork/bor/pull/160

**Comment by miguelmota** [comment]
@tilacog another block sync failure on [subgraph](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending). Thanks!

**Comment by tilacog** [comment]
@miguelmota subgraph resumed syncing.
Thanks for letting me know ;)

**Comment by miguelmota** [comment]
@tilacog there's a sync [error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) again. Sync errors seem to occur shortly after syncing is resumed

**Comment by tilacog** [comment]
Seems it is [syncing now](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending), but please let me know when it fails and I'll look into it =)

**Comment by miguelmota** [comment]
@tilacog it was making good progress but just got stuck again recently, if you don't mind taking a look. Thanks!

**Comment by miguelmota** [comment]
It's been in a [failed](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) status for a few days now 

![DeepinScreenshot_select-area_20210817095716](https://user-images.githubusercontent.com/168240/129787517-19d29c4a-c379-42e6-bc5c-551e14689234.png)

**Comment by tilacog** [comment]
@miguelmota I've just recreated the faulty block and the subgraph should be syncing now.
Thanks for the heads up!

**Comment by miguelmota** [comment]
@tilacog 🙏 looks like it synced ~800 blocks but got stuck again on block `#16698304` 😁

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16698304 (0x9135…eab9), transaction ee3110749ed21ab4102d1460ce65417a4cd75ac4aaf2b85de3477f5cc9376f76: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

**Comment by tilacog** [comment]
@miguelmota thanks for letting me know!
It is back to sync now.

**Comment by miguelmota** [comment]
@tilacog It seems to fail to sync the block every few thousand blocks and the subgraph still has 1-2M blocks to index to catch up to head. 

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16710656 (0xc5c3…6a47), transaction 78e1ceac8041c2046f732c9ed6680fc6e5ed96c6d351f873945e2af8d2c7d67a: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

Any way to automate the re-sync of the block? 

We've been trying to fully sync this subgraph for some time now 😅

**Comment by TravelingTechGuy** [comment]
Hi! Original Bug Opener™ here 😀
Encountering the same bug with this subgraph: [https://thegraph.com/legacy-explorer/subgraph/vladi-coti/cvi-polygon-usdc](https://thegraph.com/legacy-explorer/subgraph/vladi-coti/cvi-polygon-usdc).

Will re-indexing help, or is this a systemic issue?

Thanks! @azf20

**Comment by azf20** [comment]
Hey! I think @tilacog should be able to help here

**Comment by miguelmota** [comment]
[sync failure](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) `block #16735872 `

![DeepinScreenshot_select-area_20210824210243](https://user-images.githubusercontent.com/168240/130747934-76978a1a-cb3a-4a95-ae3f-b03c01c78e62.png)

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16735872 (0xbd03…dbf3), transaction e37d065810059bbba193706ee2f09f7751949f260b13ee63079efbafe09cdf01: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

**Comment by miguelmota** [comment]
[Sync error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending)

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16766208 (0xcf8f…3600), transaction 448b929a92f44df4f7e28de7144be8229485f27fa7f89c0ef165885d5f78d133: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

@tilacog

**Comment by miguelmota** [comment]
[Sync error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) - `block #16776192`

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16776192 (0x072c…c74c), transaction e2897b360222e76480388b4c55a020250717910c87c175a4e00c7230758ab4ce: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

@tilacog

**Comment by miguelmota** [comment]
[Sync error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) - `block #16785216`

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16785216 (0xdc6c…85c5), transaction 8d63b25cbc8e3fa4a62b4513d70151f8485c1fe1d5da044a83789a47219a6b55: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

@tilacog

**Comment by miguelmota** [comment]
[Sync error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) - `block #16797184`

```json
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16797184 (0x1d93…909a), transaction df952e2b5bb59beb5acbced5be788277312ec838142a808fcdbe437cff8ff12f: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

@tilacog

**Comment by shanefontaine** [comment]
Hey @tilacog 

Messages on Polygon that are sent from L1 appear as events that are emitted from the `0x0` address. For example, [here](https://polygonscan.com/tx/0x52123a30648054024aa7ecc205cc499abfed03f713ffc21be12b58a260e91d69#eventlog) is a "transaction" that has an event from one of our users that sent a message from L1 -> L2. This message emitted an event (topic `0x320958176930804eb66c2343c7343fc0367dc16249590c0f195783bee199d094`) on L2 that you can see in this transaction.

On any other chain, I would expect this event to be emitted on the transaction that actually updated state. Might this be the issue that the indexer is running into?

**Comment by miguelmota** [comment]
[Sync error](https://thegraph.com/legacy-explorer/subgraph/hop-protocol/hop-polygon?version=pending) - `block #16800192`

```json
atalError { message } nonFatalErrors {message } } }"}'
{"data":{"indexingStatusForPendingVersion":{"fatalError":{"message":"failed to process trigger: block #16800192 (0x131d…5b30), transaction 18ad7854383bfb6bf2712aa95b710ffc254ce91e882c3ed6306f929322dcdd53: Found no transaction for event"},"nonFatalErrors":[],"subgraph":"QmfDg1oHAPb7TPuy882BfBFZrJKMPNEuib98Em7rFds2nb"}}}
```

@tilacog

**Comment by schmidsi** [comment]
Hey @miguelmota and others here: We cleaned our db and apparently the new Bor version should fix our problem. Your subgraph is syncing. We hope that we now finally have a stable situation. Let us know if there are any problems.

**Comment by miguelmota** [comment]
Thanks for the update @schmidsi! Have not seen any syncing errors since the update 🎉

**Comment by schmidsi** [comment]
Nice. I close this issue then. Feel free to reopen if the problem still persists.

**Comment by schmidsi** [comment]
Seems like we have this problem now on Mumbai: https://thegraph.com/legacy-explorer/subgraph/lemonadesocial/lemonade-marketplace?version=pending

**Comment by schmidsi** [comment]
https://thegraph.com/legacy-explorer/subgraph/lemonadesocial/lemonade-marketplace?version=pending failed again with error: "failed to process trigger: block #18160704 (0xfa5e…0628), transaction ff72d1f7e52a138fa8ff82190215119118b7a66f16844eadca4d0e053cf2f6b2: Found no transaction for event"

**Comment by steve-ng** [comment]
hello @tilacog we noticed this on self-hosted graph node as well, is there an easy way to resync a block?

**Comment by tilacog** [comment]
Currently, to resync a fautly block you must delete the block record from its database table and restart graph-node afterwards.

**Comment by piavgh** [comment]
@tilacog : which table are you referring to?

**Comment by tilacog** [comment]
@piavgh It depends on how graph-node is [configured](https://github.com/graphprotocol/graph-node/blob/master/docs/config.md).

It can be either the `ethereum_blocks` table in `primary` or the `chainXX.blocks` table in a dedecated shard.

The `chainXX` is the Postgres schema identifier for a given network and can be found by querying the `public.chains` table.

**Comment by piavgh** [comment]
@tilacog : thanks, I will check to see if I can delete the error block

Btw, why the `ethereum_blocks` table is so small? I'm indexing the Cronos chain (more than 1 million block), but this table is only about 8 bytes?

![photo_2022-01-18 09 14 40](https://user-images.githubusercontent.com/4060207/149859123-115e951d-a358-4993-96b9-d2517d76fa9d.jpeg)

**Comment by vietthang207** [comment]
> @tilacog : thanks, I will check to see if I can delete the error block
> 
> Btw, why the `ethereum_blocks` table is so small? I'm indexing the Cronos chain (more than 1 million block), but this table is only about 8 bytes?
> 
> ![photo_2022-01-18 09 14 40](https://user-images.githubusercontent.com/4060207/149859123-115e951d-a358-4993-96b9-d2517d76fa9d.jpeg)

The actual table is in a different namespace. You can use \dn to list the namespace, and query the table in the other namespace.

**Comment by azf20** [comment]
Closing this for now, please re-open if you see this issue again on Polygon (Matic)

**Comment by OscBacon** [comment]
Hi, seems like Mumbai still had this issue, on block 18122880. The issue again is on a transaction from the zero address.
Note, I am using the hosted service
```
Subgraph failed with non-deterministic error: failed to process trigger: block #18122880 (0xcc82…42d3), transaction 553d80020c38eae2002b5063c264fe9b7c0b5c422859c5da9df221d40caa2f50: Found no transaction for event, retry_delay_s: 3060, attempt: 9
```

**Comment by gssvv** [comment]
Still experiencing this issue with zkSync Era mainnet:
```
Error: failed to process trigger: block #9376171 (0x01fa…404a), transaction 0000000000000000000000000000000000000000000000000000000000000000: Found no transaction for event
```

**Comment by incrypto32** [comment]
@gssvv can you share the deployment ID of the subgraph ? also that block doesnt seem to exist on zksync era

**Comment by gssvv** [comment]
@incrypto32 Sure, it's `QmW5rbKe8iq98iJD2f2D4CRqkLecu3qZpbkH7hYFvTshkp`

**Comment by KolevDarko** [comment]
I'm also facing the same thing issue, but was curious if this only happens when indexing tokens.
@gssvv are you indexing a token contract as well?

**Comment by xsteadybcgo** [comment]
The same issue with @gssvv , any workaround here?
